lib/experimental/class-wp-html-updater.php

```diff
@@ -0,0 +1,556 @@
+<?php
+
+
+class WP_HTML_Updater {
+
+	/**
+	 * @var string
+	 */
+	private $html;
+	/**
+	 * @var int
+	 */
+	private $parsed_characters = 0;
+	/**
+	 * @var string|null
+	 */
+	private $current_tag;
+	/**
+	 * @var number
+	 */
+	private $current_tag_name_ends_at;
+	/**
+	 * @var array
+	 */
+	private $current_tag_attributes = array();
+	/**
+	 * @var WP_HTML_Updater_Classnames_Set
+	 */
+	private $current_tag_class_names;
+	/**
+	 * @var array
+	 */
+	private $current_tag_touched_attributes_names = array();
+	/**
+	 * @var array
+	 */
+	private $diffs = array();
+	/**
+	 * @var null|string
+	 */
+	private $updated_html;
+
+	public function __construct( $html ) {
+		$this->html = $html;
+	}
+
+	public function find_next_tag( $tag_name, $class_name = null, $match_index = 0 ) {
+		$matched = - 1;
+		try {
+			while ( true ) {
+				// Skip through all attributes of the current tag.
+				if ( $this->current_tag ) {
+					do {
+						$attr = $this->parse_next_attribute();
+					} while ( $attr );
+					$this->after_tag();
+				}
+
+				// Match the next tag.
+				do {
+					$matches = $this->match(
+						'~<!--(?>.*?-->)|<!\[CDATA\[(?>.*?>)|<\?(?>.*?)>|<(?P<TAG_NAME>[a-z][^\x{09}\x{0a}\x{0c}\x{0d} \/>]*)~mui'
+					);
+					$this->move_parser_after_match( $matches[0] );
+				} while ( empty( $matches['TAG_NAME'][0] ) );
+
+				$this->current_tag              = $matches['TAG_NAME'][0];
+				$this->current_tag_name_ends_at = $this->index_after_match( $matches['TAG_NAME'] );
+
+				if ( $this->current_tag_matches( $tag_name, $class_name ) ) {
+					if ( ++ $matched === $match_index ) {
+						break;
+					}
+				}
+			}
+		} catch ( WP_HTML_No_Match $e ) {
+			$this->after_tag();
+			$this->parsed_characters = strlen( $this->html );
+		}
+
+		return $this;
+	}
+
+	private function parse_next_attribute() {
+		$regexp     = '~
+			# Preceeding whitespace:
+			[\x{09}\x{0a}\x{0c}\x{0d} ]*
+			(?:
+				# Either a tag end, or an attribute:
+				(?P<CLOSER>\/?>)
+				|
+				(?P<NAME>(?:
+					# Attribute names starting with an equals sign (yes, this is valid)
+					=[^=\/>\x{09}\x{0a}\x{0c}\x{0d} ]*
+					|
+					# Attribute names starting with anything other than an equals sign:
+					[^=\/>\x{09}\x{0a}\x{0c}\x{0d} ]+
+				))
+				# Optional whitespace
+				[\x{09}\x{0a}\x{0c}\x{0d} ]*
+				# Whatever terminates the attribute name
+				(?P<POST_NAME>
+					(?P<EQUALS>=)
+					[\x{09}\x{0a}\x{0c}\x{0d} ]*
+					(?:\/?>|(?P<FIRST_VALUE_CHAR>(.)))
+					|
+					\/?>
+				)?
+			)
+			~miux';
+		$name_match = $this->match( $regexp );
+
+		// No attribute, just tag closer.
+		if ( ! empty( $name_match['CLOSER'][0] ) ) {
+			return false;
+		}
```

Comment:
> In my construction I ended up taking this out because the failure to find another attribute appeared the same as finding the end of the tag. 

In my understanding, it's either:

* End of the tag – stop parsing attributes
* EOF – stop parsing attributes
* Another attribute – parse it

Which is what this code does. I may be misunderstanding what you're saying here.

> Also I ended up adding new attributes to the beginning of the list of attributes in part because it meant I didn't have to track the end of the tag, but also because if there are/were duplicates, the last one added (which I arbitrarily decided should be the "winner") would end up being the first parsed and thus the one that actually serves as the value of the attribute. Putting them at the end meant that if there were an earlier duplicate then the one we just added would be ignored.

I'm not sure I follow – that's what this PR does at the moment.

Reply:
> Which is what this code does.

This code looks for the tag opener closing `>`. All I was saying is that this step is redundant to what we need because _either_ we find another attribute (allowable within the tag opener) or we _don't_, in which case we don't care if the tag opener is closed or if nonsense follows. That is, I think if you remove the whole `(?P<CLOSER>…)` snippet nothing will change, except we don't track where the tag closer is, which we don't care about.

> I'm not sure I follow – that's what this PR does at the moment.

Yeah I think you're right: I read `current_tag_name_ends_at` as `current_tag_ends_at`

Reply:
> All I was saying is that this step is redundant to what we need because either we find another attribute (allowable within the tag opener) or we don't, in which case we don't care if the tag opener is closed or if nonsense follows.

Oh that's a great point! I just implemented that in https://github.com/WordPress/gutenberg/pull/42485/commits/3220872b494942cd4b489faaa9b008f685017213 – thanks!